aws/resource_aws_servicecatalog_product.go

```diff
@@ -0,0 +1,450 @@
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"description": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"info": {
+							Type:     schema.TypeMap,
+							Required: true,
+							Elem:     schema.TypeString,
+							ForceNew: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+							Default:  servicecatalog.ProvisioningArtifactTypeCloudFormationTemplate,
+						}, // CLOUD_FORMATION_TEMPLATE  | MARKETPLACE_AMI | MARKETPLACE_CAR
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"active": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"created_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsServiceCatalogProductCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).scconn
+	input := servicecatalog.CreateProductInput{}
+
+	if v, ok := d.GetOk("name"); ok {
+		input.Name = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("distributor"); ok {
+		input.Distributor = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("name"); ok {
+		input.Name = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("owner"); ok {
+		input.Owner = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("product_type"); ok {
+		input.ProductType = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_description"); ok {
+		input.SupportDescription = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_email"); ok {
+		input.SupportEmail = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("support_url"); ok {
+		input.SupportUrl = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		input.Tags = tagsFromMapServiceCatalog(v.(map[string]interface{}))
+	}
+
+	pa := d.Get("provisioning_artifact")
+	paList := pa.([]interface{})
+	paParameters := paList[0].(map[string]interface{})
+	artifactProperties := servicecatalog.ProvisioningArtifactProperties{}
+	artifactProperties.Description = aws.String(paParameters["description"].(string))
+	artifactProperties.Name = aws.String(paParameters["name"].(string))
+	if v, ok := paParameters["type"]; ok && v != "" {
+		artifactProperties.Type = aws.String(v.(string))
+	} else {
+		artifactProperties.Type = aws.String(servicecatalog.ProvisioningArtifactTypeCloudFormationTemplate)
+	}
+	artifactProperties.Info = make(map[string]*string)
+	for k, v := range paParameters["info"].(map[string]interface{}) {
+		artifactProperties.Info[k] = aws.String(v.(string))
+	}
+	input.IdempotencyToken = aws.String(resource.UniqueId())
+	input.SetProvisioningArtifactParameters(&artifactProperties)
```

Comment:
extract this to a seperate func similar to
https://github.com/terraform-providers/terraform-provider-aws/blob/2183d7514895383df2a08236d067f344be489a87/aws/resource_aws_service_discovery_service.go#L150

Reply:
added and using `expand`/`flatten` methods